src/Steam.Common/GameId.cs

```diff
@@ -0,0 +1,188 @@
+        /// </summary>
+        /// <param name="app">the app <see cref="App"/></param>
+        /// <param name="types">the type of the gameid <see cref="GameTypes"/>, defaults to <see cref="GameTypes.App"/></param>
+        /// <param name="modId">the id of the mod, defaults to 0x80000000</param>
+        public GameId(App app, GameTypes types = GameTypes.App, uint modId = 0x80000000)
+        {
+            _value = (ulong) modId << 32 | (ulong) types << 24 | app.Value;
+        }
+
+        /// <summary>
+        /// Checks if the GameId is a valid gameid
+        /// </summary>
+        /// <param name="id">the gameid to check the validity of</param>
+        /// <returns>Whether the GameId is valid</returns>
+        public static bool IsValid(GameId id)
+        {
+            // Todo: Maybe check if it is actualy an existing gameid against the steam API?
+            if (id._value == 0) return false;
+            if (!Enum.IsDefined(typeof(GameTypes), id._value << 24 & 0xFF)) return false;
+            
+            return true;
+        }
+
+        /// <summary>
+        /// Checks the gameIds for equality
+        /// </summary>
+        /// <param name="id">the other gameId</param>
+        /// <returns>Whether the GameIds are equal</returns>
+        public bool Equals(GameId id)
+        {
+            return GetHashCode() == id.GetHashCode();
+        }
+
+        /// <summary>
+        /// Checks for equality
+        /// </summary>
+        /// <param name="obj">expected GameId or ulong value</param>
+        /// <returns>whether or not the objects are equal</returns>
+        /// <exception cref="InvalidOperationException">If type can't be checked against</exception>
+        public override bool Equals(object obj)
+        {
+            if (obj is null) return false;
+            switch (obj)
+            {
+                case GameId gameId:
+                    return Equals(gameId);
+                case ulong numeric:
+                    return _value == numeric;
+                default:
+                    throw new InvalidOperationException($"Can't compare GameId to {obj.GetType()}");
+            }
+        }
+
+        /// <summary>
+        /// Returns a hashcode of the underlying value
+        /// </summary>
+        /// <returns>the hashcode</returns>
+        public override int GetHashCode()
+        {
+            return _value.GetHashCode();
+        }
+
+        /// <summary>
+        /// Returns a string value of the underlying value
+        /// </summary>
+        /// <returns>the string</returns>
+        public override string ToString()
+        {
+            return _value.ToString();
+        }
+
+        /// <summary>
+        /// Implicitly converts the gameid into a ulong
+        /// </summary>
+        /// <param name="id">the gameid to get the ulong value from</param>
+        /// <returns>the underlying value</returns>
+        public static implicit operator ulong(GameId id)
```

Comment:
After further consideration I don't think any of our types should have implicit cast operators. Any cast operators should be explicit.

Reply:
Other methods should also be provided for readability (i.e. `GameId.ToUInt64(gameid)`